source/server/listener_manager_impl.cc

```diff
@@ -694,20 +706,49 @@ void ListenerManagerImpl::onListenerWarmed(ListenerImpl& listener) {
   updateWarmingActiveGauges();
 }
 
-void ListenerManagerImpl::drainFilterChains(ListenerImplPtr&& listener) {
+void ListenerManagerImpl::inPlaceFilterChainUpdate(ListenerImpl& listener) {
+  auto existing_active_listener = getListenerByName(active_listeners_, listener.name());
+  auto existing_warming_listener = getListenerByName(warming_listeners_, listener.name());
+  ASSERT(existing_warming_listener != warming_listeners_.end());
+  ASSERT(*existing_warming_listener != nullptr);
+
+  (*existing_warming_listener)->debugLog("execute in place filter chain update");
+  if (existing_active_listener != active_listeners_.end()) {
+    ASSERT(*existing_active_listener != nullptr);
+    // The warmed listener should be added first so that the worker will accept new connections
+    // when it stops listening on the old listener.
+    for (const auto& worker : workers_) {
+      addListenerToWorker(*worker, listener.listenerTag(), listener, nullptr);
```

Comment:
I haven't looked at this code in a long time, but now you have multiple ListenerImpl with the same tag, right? Is that going to cause an issue anywhere? It still seems pretty strange to me that we have this structure. Can you add more comments around why that is safe along with the TODOs to merge things later?

Reply:
Based on my understanding the same tag won't cause trouble. I admit it looks less risky if tag and ListenerImpl obj is 1:1 mapping.

My understand the huge risk is not the tag, but the life time of the components in the ListenerImpl(interface ListenerConfig from workers' perspective) and refactor prior to this PR should have tackled but not well tested until this PR.

Reply:
@mattklein123 I am not sure if I understand your concern:
We planed the changes on reusing the same ListenerImpl, that is a TODO in listener_impl.cc
Still we need to notify the workers. The notifier could be build in thread local, or reuse this addListenerToWorker(). I am not sure what the comment you suggested here...